File: smoke/smokeApp/src/newSmokePPE/dataset/SimulationEnsembleDataset.js
/**
 * 
 */
function initializeGrid(coarseGrainOptions) {
    var typeName = this.type().typeName()
    var thisType = TypeRef.make({"typeName": typeName}).toType()
    if (typeof coarseGrainOptions === 'undefined') {
        coarseGrainOptions = null;
    }
    else {
      thisType.make({"id": this.id, "ensemble": this.enemble, "coarseGrainOptions": coarseGrainOptions}).merge()
    }
    // Use  TypeRef to access the bindings of this type
    var geoTimeGridType = thisType.mixins()[2].bindings()['GT'].name()
    
    //Fetch a simulation Run from the ensemble and get its' typeRef to access it's bindings
    var filter = Filter.eq("ensemble.id", "smoke_ppe_tatz")
    var simulationRunTypeName = SimulationRun.fetch({"filter": filter,"include":"id"}).objs[0].type().typeName()
    var simRun = TypeRef.make({"typeName": simulationRunTypeName}).toType()
    
    // Access the type name of the outputFile type used for this ensemble
    var outputFileTypeName = simRun.mixins()[1].bindings()['FT'].name()
    var outputFileType = TypeRef.make({"typeName": outputFileTypeName}).toType()
    
    // Get first file and call upsertGridData
    var file = outputFileType.fetch({"limit":1}).objs[0]
    file.upsertGridData(
      typeName,
      geoTimeGridType,
      this.id,
      coarseGrainOptions
    )
    return true
}

function getSimulationRunTypeName() {
  // var thisTypeName = this.type().typeName()
  // var thisType = TypeRef.make({"typeName": thisTypeName}).toType()
  var filter = Filter.eq("ensemble.id", this.ensemble.id)
  return SimulationRun.fetch({"filter": filter,"include":"id"}).objs[0].type().typeName()
}

function getSimulationOutputTypeName() {
  var thisTypeName = this.type().typeName()
  var thisType = TypeRef.make({"typeName": thisTypeName}).toType()
  return thisType.mixins()[2].bindings()['OT'].typeName()
}

function getGeoTimeGridTypeName() {
  var thisTypeName = this.type().typeName()
  var thisType = TypeRef.make({"typeName": thisTypeName}).toType()
  return thisType.mixins()[2].bindings()['GT'].typeName()
}

function getTrainedPredictionModelTypeName() {
  var thisTypeName = this.type().typeName()
  var thisType = TypeRef.make({"typeName": thisTypeName}).toType()
  return thisType.mixins()[2].bindings()['MT'].typeName()
}
function getSimulationEnsemble() {
  return SimulationEnsemble.get(this.ensemble.id)
}

function getSimulationModel() {
  var ensemble = this.getSimulationEnsemble()
  return SimulationModel.get(ensemble.model.id)
}

function getSimulationList () {
  var simulationRunTypeName = this.getSimulationRunTypeName();
  var simulationRunType = TypeRef.make({"typeName": simulationRunTypeName}).toType();
  // create list of o.simulationNumner, order by simulationNumber
  return simulationRunType.fetch({
    filter: Filter.eq("ensemble.id", this.ensemble.id)
  }).objs.map(function(o) { return o.simulationNumber }).sort(function(a, b) { return a - b });
}

function getSimulationParameterList() {
  var simulationRunTypeName = this.getSimulationRunTypeName()
  var simulationRunType = TypeRef.make({"typeName": simulationRunTypeName}).toType()
  var parameterTypeName = simulationRunType.mixins()[2].typeName()
  var fields = simulationRunType.mixins()[2].fieldTypes()

  var filteredFields = fields.filter(function(field) {
      return simulationRunType.fieldTypeDeclaredOn(field.name()).typeName() === parameterTypeName;
  });

  var fieldNames = filteredFields.map(function(field) {
      return field.name();
  });

  return fieldNames;
}

function startLoadOutputDataJob(pseudoLevelIndex, hardwareProfileId, parallelStreams, batchSize) {
  var simulationRunTypeName = this.getSimulationRunTypeName();
  var options = SimulationEnsembleRunLoadOutputDataJobOptions.make({
    typeName: simulationRunTypeName,
    datasetId: this.id,
    hardwareProfileId: hardwareProfileId,
    pseudoLevelIndex: pseudoLevelIndex, 
    parallelStreams: parallelStreams,
    batchSize: batchSize
  });
  var job = SimulationEnsembleRunLoadOutputDataJob.make({
    options: options
  }).upsert();
  job.start();
  return job;
}

function startTrainGprPredictionModelJob(gprTechnique, geoTimeGridFetchSpec, hardwareProfileId, batchSize) {
  var options = TrainGprPredictionModelJobOptions.make({
    dataset: this,
    gprTechnique: gprTechnique,
    geoTimeGridFetchSpec: geoTimeGridFetchSpec,
    hardwareProfileId: hardwareProfileId,
    batchSize: batchSize,
    X: this.getFeaturesForTechnique(gprTechnique)
  });
  var job = TrainGprPredictionModelJob.make({
    options: options
  }).upsert();
  job.start();
  return job;
}

function stageTrainedPredictionModelRowsForTechnique(geoTimeGridFetchSpec,technique, batchSize) {
  var geoGridTypeName = this.getGeoTimeGridTypeName();
  var geoGridType = TypeRef.make({"typeName": geoGridTypeName}).toType()
  // var gridPointCount = geoGridType.fetchCount(geoTimeGridFetchSpec)
  var modelTypeName = this.getTrainedPredictionModelTypeName();
  var modelType = TypeRef.make({"typeName": modelTypeName}).toType()

  var specj = geoTimeGridFetchSpec.toJson();
  specj.type = 'FetchStreamSpec'
  var streamSpec = FetchStreamSpec.make(specj)
  var gridPoints = geoGridType.fetchObjStream(streamSpec);

  var batch = [];
  var upcount = 0;
  while(gridPoints.hasNext()) {
    var gridPoint = gridPoints.next();
    // Create a composite key for the upsert
    var compositeKey = StringUtil.safeId(this.id + "_" + technique.id + "_" + gridPoint.id + "_");
    var row = modelType.make({
      "id": compositeKey,
      "dataset": this,
      "geoTimeGridPoint": gridPoint,
      "technique": technique
    });
    batch.push(row);
    if (batch.length >= batchSize || !gridPoints.hasNext()) {
      var uplist = modelType.mergeBatch(batch);
      upcount += uplist.count();
      batch = [];
    }
  }

  return upcount;
    
  }

File: smoke/smokeApp/src/newSmokePPE/dataset/SimulationEnsembleDataset.py
def getFeaturesForTechnique(this,technique):
    """
    Get the features for the GPR prediction model.
    """
    import pandas as pd
    simulationRunTypeName = this.getSimulationRunTypeName()
    featureList =  c3.PythonSerialization.deserialize(serialized=technique.serializedFeatureList)
    simulationList = c3.PythonSerialization.deserialize(serialized=technique.serializedSimulationList)

    # create filter and include spec to retrive the simulation output fields from the simulation output type
    filter = c3.Filter.inst().eq('ensemble.id',this.ensemble.id)
    # if the simulation list is not the same as the full list of available simulations, add a filter
    availableSimulations = this.getSimulationList()
    if simulationList != availableSimulations:
        filter = filter.and_().intersects('simulationNumber',simulationList)

    # convert the featureList to a string and set include to it
    include = ",".join(featureList)

    # fetch the simulation output fields
    simulationOutputFields = getattr(c3,simulationRunTypeName).fetch({
        "filter":filter,
        "include":include + ',simulationNumber'
    }).objs.toJson()

    df = pd.DataFrame(simulationOutputFields)
    df.drop(["type","id","typeIdent","meta","version"],axis=1, inplace=True)
    df.sort_values(['simulationNumber'],inplace=True)
    df.reset_index(drop=True,inplace=True)
    df.drop(['simulationNumber'],axis=1,inplace=True)
    return c3.Dataset.fromPython(df)

def getTargetForTechnique(this,technique,geoTimeGridId):
    import pandas as pd
    simulationOutputTypeName = this.getSimulationOutputTypeName()
    simulationList = c3.PythonSerialization.deserialize(serialized=technique.serializedSimulationList)

    filter = c3.Filter.inst().eq('geoTimeGridPoint',geoTimeGridId)

    availableSimulations = this.getSimulationList()
    if simulationList != availableSimulations:
        filter = filter.and_().intersects('simulationRun.simulationNumber',simulationList)

    res = getattr(c3,simulationOutputTypeName).fetch(spec={
        "filter": filter,
        "include": technique.targetName+",simulationRun.simulationNumber",
        "limit": -1
    })
    df = pd.DataFrame(res.objs.toJson())
    df['sim_num'] = df.get('simulationRun').apply(lambda x: int(x['id'].split('_')[-1]))
    df.sort_values('sim_num',inplace=True)
    df.reset_index(drop=True,inplace=True)
    df.drop(["type","id","meta","version","simulationRun","sim_num"],axis=1, inplace=True)
    return c3.Dataset.fromPython(df)
